api/appsendpoint.go

```diff
@@ -34,28 +40,44 @@ type AppsResponse struct {
 }
 
 func AppsHandler(ctx echo.Context) error{
+	const (
+		dateFormat = "20060102"
+	)
+
+	currentTime := time.Now()
+	oneWeekAgo := time.Now().AddDate(0, 0, -7)
 
-	days, err := util.IntQueryParam(ctx, "days", 1, 0, 365)
-	fmt.Println(days)
+	startDate, err := util.StringQueryParam(ctx, "startDate", oneWeekAgo.Format(dateFormat))
+	fmt.Println(startDate)
 	if err != nil {
 		return ctx.JSON(http.StatusBadRequest, ErrorResponse{Description: err.Error()})
 	}
-	db := cron.InitDB()
+
+	endDate, err := util.StringQueryParam(ctx, "endDate", currentTime.Format(dateFormat))
+	fmt.Println(endDate)
+	if err != nil{
+		return ctx.JSON(http.StatusBadRequest, ErrorResponse{Description: err.Error()})
+	}
+	eDate, err := time.Parse(dateFormat, endDate)
+	eDate.AddDate(0, 0, 1)
+	endDate = eDate.Format(dateFormat)
+
 
 	amount, err := util.IntQueryParam(ctx, "count", 10, 1, 1000)
 	fmt.Println(amount)
 	if err != nil {
 		return ctx.JSON(http.StatusBadRequest, ErrorResponse{Description: err.Error()})
 	}
 
-	apps, err := cron.GetTopApps(db, amount, days)
+	db := cron.InitDB()
```

Comment:
Since `cron.InitDB()` is already being called in the main function, it would be best if there's a way to pass it into this function instead of re-calling `InitDB` on every request. This is doubly true since all of these places you don't appear to be closing the DB connection at the ends of these functions, which could probably result in exhausting connections to the DB. It looks like this can be done with `echo` by way of passing it in when creating the handlers -- digging around I find them linking to https://echo.labstack.com/cookbook/twitter as an example of using a DB. That one's using mongodb, of course, so things will be rather different, but it might give you a good idea how to pass around a database object created in `main` rather than reconnecting to the DB in every handler like now. One thing worth noting is you may need to handle more potential errors from the DB in this case, though, particularly if the database connection gets closed or broken, you'll want the service to know how to try to reconnect.

This comment's important but the manner of doing it is a little funky so don't feel weird about asking for help. The docs on this aren't very good, heh.

Reply:
What I've ended up doing to pass around an instance of `*sql.DB` is to write a function that uses a closure to return a handler function. For example, `AppsHandler` could become `BuildAppsHandler` and it can do something like this:

``` go
func BuildAppsHandler(db *sql.DB) func(echo.Context) error {

    return func(ctx echo.Context) error {
        // Contents of current AppsHandler function without call to cron.InitDB
    }
}
```

Once you've got the method rewritten, the line in `main` becomes:

``` go
e.GET("/apps", api.BuildAppsHandler(db))
```

I don't _think_ that you need to worry about database error handling very much other than to return an error whenever you encounter a database error. The database driver should verify that the connection can be established whenever a call to `db.Connect` is made. In the case of this service, `db.Query` is being called directly, which implicitly obtains and releases a connection.

Reply:
I definitely see what you mean and will try to change it the best I can to what Sarah suggested because that makes the most sense to me. I have `defer db.Close()` in the main function; is that not enough to close the DB?

Reply:
Having the deferred `db.Close()` call should be enough to close connections associated with that instance of `sql.DB`.

Depending on how you're using database connections, you may or may not encounter problems running out of database connections. For example, if you obtain a connection by calling `db.Connection()` without calling that connection's `Close()` method then that connection will not be returned to the connection pool. doing that repeatedly will eventually exhaust the number of available connections, which will cause that `de-stats` instance to stop working. (It's also possible to cause the entire DE to stop working, depending on whether or not the exhausted connection limit is on the database host.)

The good news is that, although I haven't reviewed the code in its entirety yet, I haven't seen any cases that look like they should cause a problem. Most of the calls appear to have been made to `db.Query` directly, which will manage the database connection for you. Keep in mind that even if the connection itself is being managed for you, you'll still have to close row sets that you obtain from a database query. The examples that I've seen so far do close the row sets, though, so I don't think you'll encounter problems with that.

One concern that Ian was mentioning is that in some cases database connections can become stale or lost, forcing your program to have to drop the connection to the database completely and reestablish a new connection. I've encountered that a few times in Java, but I haven't seen that problem occur in Go yet. I think that `sql.DB` and the Postgres driver in Go do a good enough job managing the connections to shield you from problems like these. It is something good to keep in mind, though.

Hopefully that helps. As always, let us know if you have more questions.

Reply:
I'll leave another long comment, sorry @jack-mitt :laughing:

So, I commented here because I thought (incorrectly, it turns out) that this line is an example of where, in the current code without using a closure or the alternative I'll describe below, you might leak connections. In _this_ function, you're calling `cron.InitDB` again, but in this function there's no deferred close. So I was figuring that this instance of the DB would never get closed out. As the docs say, though: `The returned DB is safe for concurrent use by multiple goroutines and maintains its own pool of idle connections. Thus, the Open function should be called just once. It is rarely necessary to close a DB.`. But, also notice the "should be called just once", which is the other part and why you'll want to use a closure or my alternative, since right now you'd be opening a DB with every call to an endpoint (plus one, in the main function, that never gets used).

An alternative to using closures, which I personally prefer and is what the linked cookbook for `echo` uses, is to define a type that encapsulates the DB, and define the handlers as methods on that type. I tend to call these types `SomethingOrOtherApp` in my code. So, for example, instead of:

```
func main() {
	db := cron.InitDB()
	defer db.Close()
	e := echo.New()

       ...

	e.GET("/", api.BuildRootHandler(db))

	e.GET("/apps", api.BuildAppsHandler(db))
}
```
(which is the closure-type method as Sarah suggested)

you'd do something like:

```
// in the api namespace, not in main
type CronApp struct {
    db *sql.DB
}

// in main, as before
func main() {
	db := cron.InitDB()
	defer db.Close()
	e := echo.New()

	app := api.CronApp{db}

       ...

	e.GET("/", app.RootHandler)

	e.GET("/apps", app.AppsHandler)
}
```

And in the `api` namespace, instead of `func AppsHandler(ctx echo.Context) error{ db := ...; db.Query(...) }` like you have now, you'd do `func (c *CronApp) AppsHandler(ctx echo.Context) error { db = c.db; db.Query(...) }` or even just `c.db.Query()` directly. As I mentioned in the logging comment, you can also use this to pass around the `log` object across namespaces. You'd do something like:

```
type CronApp struct {
    db  *sql.DB
    log *logrus.Entry
}
```

and, initialize it as something like `app := api.CronApp{db: db, log: log}` (assuming you've already got a `log` variable in the main namespace as described in the other comment).

And then in functions defined as methods of `CronApp` you'd be able to do `c.log.Infof` or whatever.

Hopefully this isn't too confusing :sweat_smile: but it can make for a very nice bit of code -- it's almost, but not exactly, object-oriented this way.

One extension of this I often do is also to create a type that encapsulates the DB, and define all the database-related stuff as methods on that, or even on a type that encapsulates a transaction (mostly useful for things that write to the DB). For an example of that, you could look at [async-tasks](https://github.com/ianmcorvidae/async-tasks/blob/1ad1dd943b104fb2d859f52c6b6a0330430381b1/database.go) again -- see the `DBConnection` and `DBTx` types, and that the functions below are mostly defined as things like `func (t *DBTx) GetBaseTask(id string) (*AsyncTask, error)` and within the functions I call things like `rows, err := t.tx.Query(query, id)`. This isn't as needed as encapsulating the db itself somehow (whether by closures or a type) but it can make for some nice-to-read code in my opinion, so it's something you might consider if you like!

Reply:
I do like that option. It eliminates the extra layer of nesting, and makes it easier to pass in multiple values, as Ian mentioned.